test/asn1_encode_test.c

```diff
@@ -0,0 +1,641 @@
+    { 0, 0, 0 }
+
+/*
+ * A structure to collect all test information in.  There MUST be one instance
+ * of this for each test
+ */
+typedef int i2d_fn(void **a, unsigned char **pp);
+typedef void *d2i_fn(void **a, unsigned char **pp, long length);
+typedef void ifree_fn(void *a);
+typedef struct {
+    char *name;
+
+    /* An array of structures to compare decoded custom data with */
+    void *encode_expectations;
+    size_t encode_expectations_size;
+    size_t encode_expectations_elem_size;
+
+    /*
+     * An array of structures that are encoded into a DER blob, which is
+     * then decoded, and result gets compared with the original.
+     */
+    void *encdec_data;
+    size_t encdec_data_size;
+    size_t encdec_data_elem_size;
+
+    /* The i2d function to use with this type */
+    i2d_fn *i2d;
+    /* The d2i function to use with this type */
+    d2i_fn *d2i;
+    /* Function to free a decoded structure */
+    ifree_fn *ifree;
+} TEST_PACKAGE;
+
+/* To facilitate the creation of an encdec_data array */
+#define ENCDEC_DATA(num, znum)                  \
+    { 0xff, num, 1 }, { 0xff, 1, znum }
+#define ENCDEC_ARRAY(max, zmax)                 \
+    ENCDEC_DATA(max,zmax),                      \
+    ENCDEC_DATA(1, 1),                          \
+    ENCDEC_DATA(-1, -1),                        \
+    ENCDEC_DATA(0, ASN1_LONG_UNDEF)
+
+/***** LONG ******************************************************************/
+
+typedef struct {
+    /* If decoding is expected to succeed, set this to 1, otherwise 0 */
+    ASN1_BOOLEAN success;
+    long test_long;
+    long test_zlong;
+} ASN1_LONG_DATA;
+
+ASN1_SEQUENCE(ASN1_LONG_DATA) = {
+    ASN1_SIMPLE(ASN1_LONG_DATA, success, ASN1_FBOOLEAN),
+    ASN1_SIMPLE(ASN1_LONG_DATA, test_long, LONG),
+    ASN1_SIMPLE(ASN1_LONG_DATA, test_zlong, ZLONG)
+} static_ASN1_SEQUENCE_END(ASN1_LONG_DATA)
+
+IMPLEMENT_STATIC_ASN1_ENCODE_FUNCTIONS(ASN1_LONG_DATA)
+IMPLEMENT_STATIC_ASN1_ALLOC_FUNCTIONS(ASN1_LONG_DATA)
+
+static ASN1_LONG_DATA long_expected[] = {
+    /* The following should fail on the second because it's the default */
+    { 0xff, 0, 1 }, { 0, 0, 0 }, /* t_zero */
+    { 0, 0, 0 }, { 0xff, 1, 0x7fffffff }, /* t_longundef */
+    CUSTOM_EXPECTED_SUCCESS(1, 1), /* t_one */
+    CUSTOM_EXPECTED_FAILURE,     /* t_9bytes_1 */
+#ifdef SIXTY_FOUR_BIT_LONG
+    CUSTOM_EXPECTED_FAILURE,     /* t_8bytes_1 */
+    CUSTOM_EXPECTED_SUCCESS(LONG_MAX, LONG_MAX), /* t_8bytes_2 */
+    CUSTOM_EXPECTED_FAILURE,     /* t_8bytes_3_pad (illegal padding) */
+    CUSTOM_EXPECTED_SUCCESS(LONG_MIN, LONG_MIN), /* t_8bytes_4_neg */
+    CUSTOM_EXPECTED_FAILURE,     /* t_8bytes_5_negpad (illegal padding) */
+    CUSTOM_EXPECTED_SUCCESS(0x1ffffffff, 0x1ffffffff), /* t_5bytes_1 */
+    /*
+     * Note: because t_4bytes_1 doesn't have a negative indication, we need
+     * to code the expectation with explicit constants which are negative on
+     * 32-bit platforms and positive on larger ones.
+     */
```

Comment:
I can't quite correlate this comment with the equivalent entry for the not SIXTY_FOUR_BIT_LONG case below.